### DISCUSSION TITLE
3D coordinate and heightmap

### DISCUSSION THREAD

**Discussion by Protoxy22** [fabmax/kool#49] [Q&A] [answered]
Hi, I have a question.

I have a 2048x2048 .raw uint16 heightmap which load fine into the engine:
![image](https://github.com/fabmax/kool/assets/30320751/c1cd0953-4ea2-4d32-833a-2a1c3bc8d0c6)

But I am struggling to scale my grid heightmap to match a 
Top left point: x: -1024.0 z: 1024.0
Bottom right point: x: 1024.0 z:-1024.0

It seems that If I load my heightmap with this simple block:
```
        heightMap = HeightMap.fromRawData(Assets.loadBlobAsset("wakeisland.raw"), 7f, 2048,2048, heightOffset = 0f)

        addColorMesh {
            generate {
                grid {
                    useHeightMap(heightMap)
                }
            }
            shader = KslPbrShader {
                color { vertexColor() }
                metallic(0f)
                roughness(0.25f)
            }
        }
```

The result is a tiny 1.0f of scale heightmap, so I tried to scale(2048f) inside the grid {} to potentially have the correct 3D referential, but it's not correct.

I want to do this because I have a list of gltf models with their transformations into the world in a .json, but it doesn't match with the engine.

Have you any idea ?

Thanks

**Comment by fabmax** [comment, marked as answer]
If it's only for rendering simply setting the size within the grid block should do the trick:
```
addColorMesh {
    generate {
        grid {
            sizeX = 2048f
            sizeY = 2048f
            useHeightMap(heightMap)
        }
    }
}
```

In case you also want to use physics you can create your terrain body and then generate the mesh from the physics geometry somewhat like this:
```
    val hfGeom = HeightFieldGeometry(HeightField(heightMap, 1f, 1f))
    terrainBody = RigidStatic()
    terrainBody.attachShape(Shape(hfGeom, Physics.defaultMaterial))

    addColorMesh {
        generate {
            hfGeom.generateMesh(this)
        }
        // ... shader and stuff
    }
```
